### PR TITLE
Convert all remaining Generators to prefer statically-dimensioned Inputs and Output where possible

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: apps/HelloPyTorch/src/add_generator.cpp [hunk 1]
```diff
@@ -15,9 +15,9 @@ Func add_(const Input &input_a, const Input &input_b) {
 
 class AddGenerator : public Generator<AddGenerator> {
 public:
-    Input<Buffer<>> input_a{"input_a", 4};
-    Input<Buffer<>> input_b{"input_b", 4};
-    Output<Buffer<>> output{"output", 4};
+    Input<Buffer<void, 4>> input_a{"input_a"};
```
Comment (abadams) [inline apps/HelloPyTorch/src/add_generator.cpp:18]: This is a little unfortunate. Is it possible to make Buffer<4> work? I don't think that's doable in c++, right?

File: src/autoschedulers/adams2019/cost_model_generator.cpp [hunk 2]
```diff
@@ -15,38 +15,38 @@ using Halide::Derivative;
 // A model weight is either just an input, or an input and an output
 // (the updated weights and the ADAM state) depending on whether we're
 // doing inference or training.
-template<bool training>
```
Comment (abadams) [inline src/autoschedulers/adams2019/cost_model_generator.cpp:18]: In this file, runtime dimensionality seems much cleaner (e.g. there's value in having all the types be the same down below)

File: apps/HelloPyTorch/src/add_generator.cpp (hunk 1 above)
Comment (steven-johnson) [reply apps/HelloPyTorch/src/add_generator.cpp:18]: No, definitely not. IMHO this is a minor price to pay.

File: src/autoschedulers/adams2019/cost_model_generator.cpp (hunk 2 above)
Comment (steven-johnson) [reply src/autoschedulers/adams2019/cost_model_generator.cpp:18]: We could selectively revert this one. (I did the conversion by deleting all the old ctors temporarily and then fixing everything.)

**Comment by steven-johnson** [comment]
Tuesday Morning Review Ping

File: src/autoschedulers/adams2019/cost_model_generator.cpp (hunk 2 above)
Comment (abadams) [reply src/autoschedulers/adams2019/cost_model_generator.cpp:18]: +1 to selectively reverting this one

Comment (steven-johnson) [reply src/autoschedulers/adams2019/cost_model_generator.cpp:18]: Before I revert this, can I ask what you mean by "having all the types be the same down below"? Looking at this again, I don't know which types you mean, and so I'm confused about the specific nature of your concern. (I'm trying to understand why you feel the old approach is cleaner, so I can factor different views into future work.)

File: src/autoschedulers/adams2019/cost_model_generator.cpp [hunk 3]
```diff
@@ -448,13 +450,12 @@ class CostModel : public Generator<CostModel<training>> {
             // to the model weights.
             Derivative d_loss_d = propagate_adjoints(loss_output);
 
-            Weight *weights[] = {&head1_filter, &head1_bias,
-                                 &head2_filter, &head2_bias,
-                                 &filter1, &bias1};
-
-            for (Weight *w : weights) {
```
Comment (abadams) [inline src/autoschedulers/adams2019/cost_model_generator.cpp:455]: It's losing the ability to iterate over the weights here that I was referring to.

Comment (steven-johnson) [reply src/autoschedulers/adams2019/cost_model_generator.cpp:455]: That could be fixed pretty trivially by making `backprop()` a virtual method and adding a `ModelWeightBase` class.

Comment (steven-johnson) [reply src/autoschedulers/adams2019/cost_model_generator.cpp:455]: (Just pushed a revision to do this.)

**Pull request closed without merging**